ext/standard/streamsfuncs.c

```diff
@@ -756,115 +782,132 @@ static int stream_array_emulate_read_fd_set(zval *stream_array)
-			RETURN_THROWS();
-		}
-	}
+    if (secnull && !usecnull) {
+        if (usec != 0) {
+            zend_argument_value_error(5, "must be null when argument #4 ($seconds) is null");
+            FD_BIGSET_FREE(&rfds);
+            FD_BIGSET_FREE(&wfds);
+            FD_BIGSET_FREE(&efds);
+            RETURN_THROWS();
+        }
+    }
 
-	/* If seconds is not set to null, build the timeval, else we wait indefinitely */
-	if (!secnull) {
-		if (sec < 0) {
-			zend_argument_value_error(4, "must be greater than or equal to 0");
-			RETURN_THROWS();
-		} else if (usec < 0) {
-			zend_argument_value_error(5, "must be greater than or equal to 0");
-			RETURN_THROWS();
-		}
+    if (!secnull) {
+        if (sec < 0) {
+            zend_argument_value_error(4, "must be greater than or equal to 0");
+            FD_BIGSET_FREE(&rfds);
+            FD_BIGSET_FREE(&wfds);
+            FD_BIGSET_FREE(&efds);
+            RETURN_THROWS();
+        } else if (usec < 0) {
+            zend_argument_value_error(5, "must be greater than or equal to 0");
+            FD_BIGSET_FREE(&rfds);
+            FD_BIGSET_FREE(&wfds);
+            FD_BIGSET_FREE(&efds);
+            RETURN_THROWS();
+        }
+
+        tv.tv_sec = (long)(sec + (usec / 1000000));
+        tv.tv_usec = (long)(usec % 1000000);
+        tv_p = &tv;
+    }
 
-		/* Windows, Solaris and BSD do not like microsecond values which are >= 1 sec */
-		tv.tv_sec = (long)(sec + (usec / 1000000));
-		tv.tv_usec = (long)(usec % 1000000);
-		tv_p = &tv;
-	}
-
-	/* slight hack to support buffered data; if there is data sitting in the
-	 * read buffer of any of the streams in the read array, let's pretend
-	 * that we selected, but return only the readable sockets */
-	if (r_array != NULL) {
-		retval = stream_array_emulate_read_fd_set(r_array);
-		if (retval > 0) {
-			if (w_array != NULL) {
-				zval_ptr_dtor(w_array);
-				ZVAL_EMPTY_ARRAY(w_array);
-			}
-			if (e_array != NULL) {
-				zval_ptr_dtor(e_array);
-				ZVAL_EMPTY_ARRAY(e_array);
-			}
-			RETURN_LONG(retval);
-		}
-	}
+    // Support for buffered data
+    if (r_array != NULL) {
+        retval = stream_array_emulate_read_fd_set(r_array);
+        if (retval > 0) {
+            if (w_array != NULL) {
+                zval_ptr_dtor(w_array);
+                ZVAL_EMPTY_ARRAY(w_array);
+            }
+            if (e_array != NULL) {
+                zval_ptr_dtor(e_array);
+                ZVAL_EMPTY_ARRAY(e_array);
+            }
+            FD_BIGSET_FREE(&rfds);
+            FD_BIGSET_FREE(&wfds);
+            FD_BIGSET_FREE(&efds);
+            RETURN_LONG(retval);
+        }
+    }
 
-	retval = php_select(max_fd+1, &rfds, &wfds, &efds, tv_p);
+	#ifdef __USE_XOPEN
+		retval = php_select(max_fd + 1, (fd_set *)rfds.fds_bits, (fd_set *)wfds.fds_bits, (fd_set *)efds.fds_bits, tv_p);
+	#else
+		retval = php_select(max_fd + 1, (fd_set *)rfds.__fds_bits, (fd_set *)wfds.__fds_bits, (fd_set *)efds.__fds_bits, tv_p);
```

Comment:
We could define an accessor for `__fds_bits` to avoid the ifdef here (and also above the `FD_BIGSET_ZERO` definition earlier in this file) : 

```
#define FD_BIGSET_BITS(set) ((set)->__fd_bits)

retval = php_select(max_fd + 1, (fd_set *)FD_BIGSET_BITS(rfds), ..., ..., tv_p);
```

Reply:
Thank you for looking over the code. Since this is my first PR for php-src I tried to get info about standards before sending it. I'll try to do better next time.

Am I supposed to update my PR based on your feedback here, or is that something others do?

Reply:
Thank you for this PR :) 

Yes you can update your branch directly. It will be merged when there are no more feedbacks or change requests, and if @bukka approves it.

Reply:
@arnaud-lb I have updated my branch taking in your feedback, except for the accessor for FD_BIGSET_BITS, which I tried but it didn't work and my C skills aren't there yet.